packages/client/src/stats/SfuStatsReporter.ts

```diff
@@ -20,34 +28,86 @@ export class SfuStatsReporter {
   private readonly sfuClient: StreamSfuClient;
   private readonly subscriber: Subscriber;
   private readonly publisher?: Publisher;
+  private readonly microphone: MicrophoneManager;
+  private readonly camera: CameraManager;
+  private readonly state: CallState;
 
   private intervalId: NodeJS.Timeout | undefined;
+  private subscriptions: Array<() => void> = [];
   private readonly sdkName: string;
   private readonly sdkVersion: string;
   private readonly webRTCVersion: string;
+  private readonly inputDevices = new Map<'mic' | 'camera', InputDevices>();
 
   constructor(
     sfuClient: StreamSfuClient,
-    { options, clientDetails, subscriber, publisher }: SfuStatsReporterOptions,
+    {
+      options,
+      clientDetails,
+      subscriber,
+      publisher,
+      microphone,
+      camera,
+      state,
+    }: SfuStatsReporterOptions,
   ) {
     this.sfuClient = sfuClient;
     this.options = options;
     this.subscriber = subscriber;
     this.publisher = publisher;
-    const webRTCInfo = getWebRTCInfo();
+    this.microphone = microphone;
+    this.camera = camera;
+    this.state = state;
 
     const { sdk, browser } = clientDetails;
-
     this.sdkName = getSdkName(sdk);
     this.sdkVersion = getSdkVersion(sdk);
 
     // The WebRTC version if passed from the SDK, it is taken else the browser info is sent.
+    const webRTCInfo = getWebRTCInfo();
     this.webRTCVersion =
       webRTCInfo?.version ||
       `${browser?.name || ''}-${browser?.version || ''}` ||
       'N/A';
   }
 
+  private observeDevice = (
+    device: CameraManager | MicrophoneManager,
+    kind: 'mic' | 'camera',
+  ) => {
+    const { hasBrowserPermission$ } = device.state;
+    const permissionsSubscription = createSubscription(
+      combineLatest([hasBrowserPermission$, this.state.ownCapabilities$]),
+      ([hasPermission, ownCapabilities]) => {
+        const hasCapability =
+          kind === 'mic'
+            ? ownCapabilities.includes(OwnCapability.SEND_AUDIO)
+            : ownCapabilities.includes(OwnCapability.SEND_VIDEO);
+        if (!hasPermission || !hasCapability) {
+          this.inputDevices.set(kind, {
+            currentDevice: '',
+            availableDevices: [],
+            isPermitted: false,
+          });
+          return;
+        }
+        const listDevicesSubscription = createSubscription(
```

Comment:
I'm a bit worried about creating a subscription inside a subscription, especially since we won't clean up "inner" subscriptions until `stop` is called.

Maybe we can use something like `iif` to subscribe to `listDevices()` only if we have both permission and capabilities?

Reply:
yep, it can happen. rarely, but it can still happen. let me try to fix this.

Reply:
@myandrienko please check https://github.com/GetStream/stream-video-js/pull/1533/commits/b2b6e6cee9ffc066524c7df3969c7623f872ac66